File: app/integrations/unleashed_client.py
"""
Unleashed Software API Client
Handles authentication and communication with the Unleashed API
"""

import hashlib
import hmac
import requests
import os
from datetime import datetime
from urllib.parse import urlparse, parse_qs
from dotenv import load_dotenv

load_dotenv('.env.local')


class UnleashedClient:
    """Client for interacting with Unleashed Software API"""
    
    def __init__(self):
        self.api_id = os.getenv('UNLEASHED_API_ID')
        self.api_key = os.getenv('UNLEASHED_API_KEY')
        self.base_url = os.getenv('UNLEASHED_API_URL', 'https://api.unleashedsoftware.com')
        
        if not self.api_id or not self.api_key:
            raise ValueError("Unleashed API credentials not found in environment variables")
    
    def _generate_signature(self, query_string=""):
        """
        Generate HMAC-SHA256 signature for API authentication
        
        Args:
            query_string: The query string part of the URL (if any)
        
        Returns:
            Base64 encoded signature
        """
        if query_string:
            # Remove leading '?' if present
            if query_string.startswith('?'):
                query_string = query_string[1:]
        
        # Create the signature using HMAC-SHA256
        signature = hmac.new(
            self.api_key.encode('utf-8'),
            query_string.encode('utf-8'),
            hashlib.sha256
        ).digest()
        
        # Return base64 encoded signature
        import base64
        return base64.b64encode(signature).decode('utf-8')
    
    def _make_request(self, endpoint, method='GET', params=None, data=None):
        """
        Make authenticated request to Unleashed API
        
        Args:
            endpoint: API endpoint (e.g., '/Customers')
            method: HTTP method (GET, POST, PUT, DELETE)
            params: Query parameters
            data: Request body data
        
        Returns:
            Response JSON data
        """
        # Build full URL
        url = f"{self.base_url}{endpoint}"
        
        # Build query string for signature
        query_string = ""
        if params:
            query_pairs = []
            for key, value in sorted(params.items()):
                query_pairs.append(f"{key}={value}")
            query_string = "&".join(query_pairs)
        
        # Generate signature
        signature = self._generate_signature(query_string)
        
        # Set headers
        headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'api-auth-id': self.api_id,
            'api-auth-signature': signature
        }
        
        # Make request
        try:
            if method == 'GET':
                response = requests.get(url, headers=headers, params=params)
            elif method == 'POST':
                response = requests.post(url, headers=headers, params=params, json=data)
            elif method == 'PUT':
                response = requests.put(url, headers=headers, params=params, json=data)
            elif method == 'DELETE':
                response = requests.delete(url, headers=headers, params=params)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            print(f"API request failed: {e}")
            if hasattr(e, 'response') and e.response is not None:
                print(f"Response status: {e.response.status_code}")
                print(f"Response body: {e.response.text}")
            raise
    
    # Product Methods
    def get_products(self, page=1, page_size=200):
        """Get products from Unleashed"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/Products', params=params)
    
    def get_product(self, product_guid):
        """Get a specific product by GUID"""
        return self._make_request(f'/Products/{product_guid}')
    
    # Stock on Hand Methods
    def get_stock_on_hand(self, page=1, page_size=200):
        """Get current stock on hand"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/StockOnHand', params=params)
    
    # Sales Order Methods
    def get_sales_orders(self, page=1, page_size=200, order_status=None):
        """Get sales orders"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        if order_status:
            params['orderStatus'] = order_status
        return self._make_request('/SalesOrders', params=params)
    
    def get_sales_order(self, order_guid):
        """Get a specific sales order by GUID"""
        return self._make_request(f'/SalesOrders/{order_guid}')
    
    # Purchase Order Methods
    def get_purchase_orders(self, page=1, page_size=200):
        """Get purchase orders"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/PurchaseOrders', params=params)
    
    # Customer Methods
    def get_customers(self, page=1, page_size=200):
        """Get customers"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/Customers', params=params)
    
    def get_customer(self, customer_guid):
        """Get a specific customer by GUID"""
        return self._make_request(f'/Customers/{customer_guid}')
    
    # Supplier Methods
    def get_suppliers(self, page=1, page_size=200):
        """Get suppliers"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/Suppliers', params=params)
    
    # Warehouse Methods
    def get_warehouses(self):
        """Get all warehouses"""
        return self._make_request('/Warehouses')
    
    # Bill of Materials Methods
    def get_bill_of_materials(self, page=1, page_size=200):
        """Get bills of materials"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/BillsOfMaterials', params=params)
    
    # Stock Adjustments
    def get_stock_adjustments(self, page=1, page_size=200):
        """Get stock adjustments"""
        params = {
            'pageNumber': page,
            'pageSize': page_size
        }
        return self._make_request('/StockAdjustments', params=params)
    
    # Test connection
    def test_connection(self):
        """Test API connection by fetching warehouses"""
        try:
            result = self.get_warehouses()
            return True, f"Successfully connected. Found {len(result.get('Items', []))} warehouses."
        except Exception as e:
            return False, f"Connection failed: {str(e)}"


if __name__ == "__main__":
    # Test the client
    try:
        client = UnleashedClient()
        success, message = client.test_connection()
        print(f"Connection test: {message}")
        
        if success:
            # Try to get some products
            print("\nFetching products...")
            products = client.get_products(page_size=5)
            
            if 'Items' in products:
                print(f"Found {products.get('Total', 0)} products total")
                for product in products['Items'][:5]:
                    print(f"  - {product.get('ProductCode', 'N/A')}: {product.get('ProductDescription', 'N/A')}")
            
            # Try to get stock on hand
            print("\nFetching stock on hand...")
            stock = client.get_stock_on_hand(page_size=5)
            
            if 'Items' in stock:
                print(f"Found {stock.get('Total', 0)} stock items total")
                for item in stock['Items'][:5]:
                    print(f"  - {item.get('ProductCode', 'N/A')}: {item.get('QtyOnHand', 0)} units")
    
    except Exception as e:
        print(f"Error: {e}")

File: services/unleashed/UnleashedClient.js
import crypto from 'crypto';
import axios from 'axios';
import { logInfo, logError, logWarn } from '../observability/structuredLogger.js';

class UnleashedClient {
  constructor() {
    this.apiId = process.env.UNLEASHED_API_ID || '';
    this.apiKey = process.env.UNLEASHED_API_KEY || '';
    this.baseUrl = 'https://api.unleashedsoftware.com';
    this.timeout = 30000; // 30 seconds
    this.retryCount = 3;
    this.retryDelay = 1000; // 1 second

    // Validate credentials
    if (!this.apiId || !this.apiKey) {
      logWarn('Unleashed API credentials not configured', {
        hasApiId: !!this.apiId,
        hasApiKey: !!this.apiKey
      });
    }
  }

  /**
   * Generate HMAC-SHA256 signature for API authentication
   */
  getSignature(query) {
    try {
      const hmac = crypto.createHmac('sha256', this.apiKey);
      hmac.update(query || '');
      return hmac.digest('base64');
    } catch (error) {
      logError('Failed to generate Unleashed signature', error);
      throw new Error('Authentication signature generation failed');
    }
  }

  /**
   * Make authenticated request to Unleashed API
   */
  async request(endpoint, params = {}, method = 'GET') {
    if (!this.apiId || !this.apiKey) {
      throw new Error('Unleashed API credentials not configured');
    }

    const query = new URLSearchParams(params).toString();
    const signature = this.getSignature(query);

    const config = {
      method,
      url: `${this.baseUrl}${endpoint}`,
      params: method === 'GET' ? params : undefined,
      data: method !== 'GET' ? params : undefined,
      headers: {
        'api-auth-id': this.apiId,
        'api-auth-signature': signature,
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      timeout: this.timeout
    };

    let lastError;
    for (let attempt = 1; attempt <= this.retryCount; attempt++) {
      try {
        logInfo('Unleashed API request', {
          endpoint,
          method,
          attempt,
          hasParams: Object.keys(params).length > 0
        });

        const response = await axios(config);

        logInfo('Unleashed API response', {
          endpoint,
          status: response.status,
          itemCount: response.data?.Items?.length || 0
        });

        return response.data;
      } catch (error) {
        lastError = error;

        if (error.response) {
          // Server responded with error
          logError('Unleashed API error response', {
            endpoint,
            status: error.response.status,
            message: error.response.data?.Message || error.message,
            attempt
          });

          // Don't retry client errors (4xx)
          if (error.response.status >= 400 && error.response.status < 500) {
            throw new Error(`Unleashed API error: ${error.response.data?.Message || error.message}`);
          }
        } else if (error.request) {
          // Request made but no response
          logError('Unleashed API no response', {
            endpoint,
            attempt,
            error: error.message
          });
        } else {
          // Error in request setup
          logError('Unleashed API request setup error', {
            endpoint,
            attempt,
            error: error.message
          });
        }

        // Wait before retry
        if (attempt < this.retryCount) {
          await new Promise(resolve => setTimeout(resolve, this.retryDelay * attempt));
        }
      }
    }

    throw new Error(`Unleashed API request failed after ${this.retryCount} attempts: ${lastError?.message}`);
  }

  /**
   * Get all products from Unleashed
   */
  async getProducts(page = 1, pageSize = 200) {
    try {
      const response = await this.request('/Products', {
        page,
        pageSize
      });

      return {
        items: response.Items || [],
        pagination: response.Pagination || {},
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch Unleashed products', error);
      throw error;
    }
  }

  /**
   * Get stock on hand for all products
   */
  async getStockOnHand(page = 1, pageSize = 200) {
    try {
      const response = await this.request('/StockOnHand', {
        page,
        pageSize
      });

      return {
        items: response.Items || [],
        pagination: response.Pagination || {},
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch stock on hand', error);
      throw error;
    }
  }

  /**
   * Get purchase orders
   */
  async getPurchaseOrders(page = 1, pageSize = 200, status = null) {
    try {
      const params = { page, pageSize };
      if (status) params.orderStatus = status;

      const response = await this.request('/PurchaseOrders', params);

      return {
        items: response.Items || [],
        pagination: response.Pagination || {},
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch purchase orders', error);
      throw error;
    }
  }

  /**
   * Get sales orders
   */
  async getSalesOrders(page = 1, pageSize = 200, status = null) {
    try {
      const params = { page, pageSize };
      if (status) params.orderStatus = status;

      const response = await this.request('/SalesOrders', params);

      return {
        items: response.Items || [],
        pagination: response.Pagination || {},
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch sales orders', error);
      throw error;
    }
  }

  /**
   * Get stock movements
   */
  async getStockMovements(productCode = null, startDate = null, endDate = null) {
    try {
      const params = {};
      if (productCode) params.productCode = productCode;
      if (startDate) params.startDate = startDate;
      if (endDate) params.endDate = endDate;

      const response = await this.request('/StockMovements', params);

      return {
        items: response.Items || [],
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch stock movements', error);
      throw error;
    }
  }

  /**
   * Get warehouses
   */
  async getWarehouses() {
    try {
      const response = await this.request('/Warehouses');

      return {
        items: response.Items || [],
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch warehouses', error);
      throw error;
    }
  }

  /**
   * Get stock adjustments
   */
  async getStockAdjustments(page = 1, pageSize = 200) {
    try {
      const response = await this.request('/StockAdjustments', {
        page,
        pageSize
      });

      return {
        items: response.Items || [],
        pagination: response.Pagination || {},
        total: response.Total || 0
      };
    } catch (error) {
      logError('Failed to fetch stock adjustments', error);
      throw error;
    }
  }

  /**
   * Test connection to Unleashed API
   */
  async testConnection() {
    try {
      const response = await this.request('/Currencies', { page: 1, pageSize: 1 });
      return {
        success: true,
        message: 'Successfully connected to Unleashed API'
      };
    } catch (error) {
      return {
        success: false,
        message: `Connection failed: ${error.message}`
      };
    }
  }
}

// Singleton instance
let clientInstance = null;

export function getUnleashedClient() {
  if (!clientInstance) {
    clientInstance = new UnleashedClient();
  }
  return clientInstance;
}

export default UnleashedClient;